### PR TITLE
Switch jsxstyle and jsxstyle-utils to TypeScript

### PR DESCRIPTION
**Pull request by meyer**
This is a first pass at switching jsxstyle to TypeScript. Main motivation is that maintaining separate type definition files seems like it would lead to inconsistencies.

@appsforartists and @steadicat, I’d love some input if you’ve got the time.

### CODE REVIEW

File: packages/jsxstyle-utils/src/componentStyles.ts [hunk 1]
```diff
@@ -1,3 +1,7 @@
+export interface StyleObject {
+  [key: string]: { [key: string]: string | number } | null;
```
Comment (steadicat) [inline packages/jsxstyle-utils/src/componentStyles.ts]: For our project, I lifted [CSS types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L817) from `@types/react`. It's pretty important for `jsxstyle` to have strict types for CSS props. Otherwise any prop on any component goes, which would hide a ton of errors.

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 2]
```diff
@@ -2,6 +2,26 @@ import getStyleKeysForProps from './getStyleKeysForProps';
 import addStyleToHead from './addStyleToHead';
 import stringHash from './stringHash';
 
+export interface InsertRuleCallback {
+  (rule: string, props?: {}): boolean | void;
+}
+
+export interface StyleCache {
+  reset: { (): void };
```
Comment (steadicat) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: Is this the same as `reset(): void`?

**Comment by jaredpalmer** [comment]
Awwww yeah

File: packages/jsxstyle/jsxstyle.tsx [hunk 3]
```diff
@@ -1,21 +1,30 @@
...
+}
+
+function factory(
+  displayName: string,
+  defaultProps?: { [key: string]: React.ReactText } | null
+) {
+  const tagName = 'div';
+
+  return class JsxstyleComponent extends React.Component<
+    JsxstyleProps & { [key: string]: any }
```
Comment (steadicat) [inline packages/jsxstyle/jsxstyle.tsx]: I highly recommend removing `{[key: string]: any}` here, for the reason mentioned above.

File: packages/jsxstyle/jsxstyle.tsx [hunk 4]
```diff
@@ -1,21 +1,30 @@
...
-    process.env.NODE_ENV === 'development' &&
-    (typeof displayName !== 'string' || !displayName)
-  ) {
-    throw new Error(
-      'makeReactStyleComponentClass expects param 1 to be a valid displayName'
-    );
-  }
+export interface JsxstyleProps {
+  className?: string;
+  component?: string | React.ComponentClass | React.SFC;
```
Comment (steadicat) [inline packages/jsxstyle/jsxstyle.tsx]: Use `keyof JSX.IntrinsicElements` instead of `string`.

File: packages/jsxstyle/jsxstyle.tsx [hunk 5]
```diff
@@ -1,21 +1,30 @@
...
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
```
Comment (steadicat) [inline packages/jsxstyle/jsxstyle.tsx]: Use `React.AllHTMLAttributes<T> & React.ClassAttributes<T>`.

File: packages/jsxstyle/jsxstyle.tsx [hunk 6]
```diff
@@ -1,21 +1,30 @@
...
-    (typeof displayName !== 'string' || !displayName)
-  ) {
-    throw new Error(
-      'makeReactStyleComponentClass expects param 1 to be a valid displayName'
-    );
-  }
+export interface JsxstyleProps {
+  className?: string;
+  component?: string | React.ComponentClass | React.SFC;
+  mediaQueries?: { [key: string]: string };
```
Comment (steadicat) [inline packages/jsxstyle/jsxstyle.tsx]: Enumerate all possible media queries here if possible.

File: packages/jsxstyle-utils/src/componentStyles.ts [hunk 7]
```diff
@@ -6,4 +10,4 @@ export default {
   Row: { display: 'flex', flexDirection: 'row' },
   Col: { display: 'flex', flexDirection: 'column' },
   Grid: { display: 'grid' },
-};
+} as StyleObject;
```
Comment (steadicat) [inline packages/jsxstyle-utils/src/componentStyles.ts]: Why is this needed? TS should be able to infer a better type for this.

File: packages/jsxstyle-utils/src/addStyleToHead.ts [hunk 8]
```diff
@@ -23,18 +23,19 @@ if (canUseDOM && !styleElement) {
   document.head.appendChild(styleElement);
 }
 
-export default function addStyleToHead(rule) {
-  if (canUseDOM) {
+export default function addStyleToHead(rule: string) {
+  if (styleElement) {
+    const sheet = styleElement.sheet as CSSStyleSheet;
```
Comment (steadicat) [inline packages/jsxstyle-utils/src/addStyleToHead.ts]: Why is the cast needed here?

**Comment by steadicat** [comment]
BTW `React.CSSProperties` has a string index accessor, which sucks. I forked it and removed it. I would also replace all the `| any` with actual allowed values.

File: packages/jsxstyle-utils/src/addStyleToHead.ts (hunk 8 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/addStyleToHead.ts]: @steadicat the cast feels gross but without it I get `Property 'insertRule' does not exist on type 'StyleSheet'`. Apparently (according to https://github.com/Microsoft/TypeScript/issues/15319) this issue occurs because `styleElement.sheet` might be one of several types.

File: packages/jsxstyle-utils/src/componentStyles.ts (hunk 7 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/componentStyles.ts]: @steadicat good call, I’ll remove it.

File: packages/jsxstyle/jsxstyle.tsx (hunk 6 above)
Comment (meyer) [reply packages/jsxstyle/jsxstyle.tsx]: The main problem here is that the media query key can be anything the user specifies. The API does what it needs to do but I’m not super happy with how open-ended it makes things.

File: packages/jsxstyle/jsxstyle.tsx (hunk 4 above)
Comment (meyer) [reply packages/jsxstyle/jsxstyle.tsx]: 🙏 thank you

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 2 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: @steadicat heh, yeah. can you tell i’m new to TypeScript? 😅

**Comment by meyer** [comment]
@steadicat I’m looking into a way of auto-generating `CSSProperties` with jsxstyle-specific additions, like the pseudoclass and pseudoelement prefixes. If you’ve got ideas let me know! For now I’m probably going to parse the relevant parts of `@types/react` with babylon and add the prefixed style props at the AST level.

The main issue (as I mentioned above) is that the media query prefix is user-specified and can therefore be _anything_. I’m not sure what to do about that type-wise other than to either use an index accessor (😩) or rethink the media query feature entirely.

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 9]
```diff
@@ -14,17 +34,17 @@ function alreadyInjected() {
   );
 }
 
-function getStringHash(key) {
+function getStringHash(key: string, _: {}) {
```
Comment (meyer) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: Is there an official way to ignore a second parameter? It’s passed to `getClassNameForKey` on line 72 and I get why TS would complain about function signature being different, but an unused underscore param seems hella nasty. It gets minified out so I’m not too stressed about it, but just seems like there’s got to be a better way to handle this.

File: packages/jsxstyle-utils/src/addStyleToHead.ts (hunk 8 above)
Comment (steadicat) [reply packages/jsxstyle-utils/src/addStyleToHead.ts]: That’s weird, the [type definition of LinkStyle.sheet](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L13685) only has one type.

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 9 above)
Comment (jaredpalmer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Yeah prefix name with underscore and it will be ignored if unused

Comment (steadicat) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: You can use function overloads for this:
```ts
function getStringHash(key: string, props: any);
function getStringHash(key: string) {
  ...
}
```

File: packages/jsxstyle-utils/src/addStyleToHead.ts (hunk 8 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/addStyleToHead.ts]: Yeah, it’s funky. There’s currently only one type of `StyleSheet` in the spec (`CSSStyleSheet`) but there might be more than one in the future.

https://github.com/Microsoft/TypeScript/issues/15319#issuecomment-296584050
https://drafts.csswg.org/cssom/#the-stylesheet-interface

**Comment by meyer** [comment]
I added a script a little while ago that extracts out `CSSProperties` from `@types/react`, and I just added prefixing to the script locally. 59 prefixes are added to each prop in the `CSSProperties` interface. The generated file is 23K lines—a few bytes short of a megabyte (!!!).

image

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 10]
```diff
@@ -2,6 +2,20 @@ import getStyleKeysForProps from './getStyleKeysForProps';
 import addStyleToHead from './addStyleToHead';
 import stringHash from './stringHash';
 
+export interface InsertRuleCallback {
+  (rule: string, props?: {}): boolean | void;
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: I feel like you could get fancy here and type `props` more tightly, the same way that `document.createElement` returns a different type based on which string you pass it.  I think it's something like:


```typescript
export interface InsertRuleCallback<K extends keyof RuleToProps> {
  (rule: K, props?: RuleToProps[K]): boolean | void;
}
```

and it looks like this is just a function type.  I'd have expected

```ts
export type InsertRuleCallback<K extends keyof RuleToProps> = (rule: K, props?: RuleToProps[K]) => boolean | void;
```
but that's just a style thing.  I've never seen `interface` used to represent a single function signature before.

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 11]
```diff
@@ -2,6 +2,20 @@ import getStyleKeysForProps from './getStyleKeysForProps';
...
+}
+
+export interface StyleCache {
+  reset(): void;
+  getClassName(props: { [key: string]: any }, classNameProp?: string): string;
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: In my codebases, I like to have a `Dict` type:

```ts
export type Dict<T> = {
  [key: string]: T,
};
```
which makes this a little easier to express:

```ts
  getClassName(props: Dict<any>, classNameProp?: string): string;
```

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 12]
```diff
@@ -14,17 +28,17 @@ function alreadyInjected() {
   );
 }
 
-function getStringHash(key) {
+function getStringHash(key: string, props: {}) {
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: You're not using `props` here.  I take it this is because you have call sites that pass two params.

The 'correct' way to express this is probably something like:

```ts
export type GetClassNameForKey = (key: string, props: {}) => string;

let getClassNameForKey = getStringHash as GetClassNameForKey;
```

but if you wanted something quicker and more concise, you could make `props` optional:

```typescript
function getStringHash(key: string, props?: {}): string {
```

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 13]
```diff
@@ -14,17 +28,17 @@ function alreadyInjected() {
...
 
 export default function getStyleCache() {
   let _classNameCache = {};
   let getClassNameForKey = getStringHash;
-  let onInsertRule = null;
+  let onInsertRule: InsertRuleCallback;
   let pretty = false;
 
-  const styleCache = {};
+  const styleCache: StyleCache = <any>{};
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: `({} as any)` is easier to read.  I don't often see `<>` used for casting.

But more holistically, I don't think you need to cast this at all - it's just sidestepping the type checker.  Can you just refactor it to

```typescript
const styleCache: StyleCache = {
  reset() {
    _classNameCache = {}
  }
  // ...
}
```

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts [hunk 14]
```diff
@@ -35,7 +35,7 @@ const specialCaseProps = {
   mediaQueries: true,
 };
 
-export default function getStyleKeysForProps(props, pretty = false) {
+export default function getStyleKeysForProps(props: any, pretty = false) {
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: I wonder again if there is a tighter way to type `props`.

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts [hunk 15]
```diff
@@ -52,7 +52,14 @@ export default function getStyleKeysForProps(props, pretty = false) {
   let usesMediaQueries = false;
 
   // return value
-  const styleKeyObj = {};
+  const styleKeyObj: {
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: You should extract this type and explicitly use it as the return value on the function.  That will make the function signature easier to skim, and keep you from having to cast an empty dictionary to match a type it doesn't yet implement.

File: packages/jsxstyle/jsxstyle.tsx [hunk 16]
```diff
@@ -1,21 +1,41 @@
...
+  className?: string;
+  component?: ComponentProp;
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
+  style?: CSSProperties;
+}
+
+function factory(
+  displayName: string,
+  defaultProps?: { [key: string]: React.ReactText } | null
```
Comment (appsforartists) [inline packages/jsxstyle/jsxstyle.tsx]: This is why `Dict` is cool. 😄

File: rollup.config.js [hunk 17]
```diff
@@ -11,50 +11,25 @@ const external = [
...
-          // don't need fancy typeof
-          'transform-es2015-typeof-symbol',
-        ],
-      },
-    ],
-  ],
-  plugins: [
-    require.resolve('babel-plugin-transform-class-properties'),
-    require.resolve('babel-plugin-transform-object-assign'),
-    require.resolve('babel-plugin-external-helpers'),
```
Comment (appsforartists) [inline rollup.config.js]: Babel is awesome, but it's nice to not have to worry about which not-yet-standardized features you're opting into.

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 10 above)
Comment (steadicat) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Agree that this should probably be:
```ts
function InsertRuleCallback(rule: string, props?: {}): boolean | void;
```

File: packages/jsxstyle/jsxstyle.tsx [hunk 18]
```diff
@@ -1,21 +1,41 @@
...
+export type ComponentProp =
+  | keyof JSX.IntrinsicElements
+  | React.ComponentClass<ComponentPropProps>
+  | React.SFC<ComponentPropProps>;
+
+export interface JsxstyleProps extends CSSProperties {
+  className?: string;
+  component?: ComponentProp;
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
```
Comment (steadicat) [inline packages/jsxstyle/jsxstyle.tsx]: `props` on the public interface really shouldn't be a `{[key: string]: any}`. How about something like this:
```ts
export type AnyComponent<ComponentProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<ComponentProps>
  | React.SFC<ComponentProps>;

export type JsxstyleProps<ComponentProps> = {
  className?: string;
  component?: AnyComponent<ComponentProps>;
  mediaQueries?: {[key: string]: string};
  props?: ComponentProps;
  style?: CSSProperties;
} & CSSProperties;
```

File: packages/jsxstyle/preact/jsxstyle-preact.tsx [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+export type ComponentProp =
+  | keyof JSX.IntrinsicElements
+  | preact.AnyComponent<any, any>
+  | any;
+
+export interface JsxstyleProps extends CSSProperties {
+  className?: string;
+  component?: keyof JSX.IntrinsicElements | React.ComponentClass | React.SFC;
```
Comment (steadicat) [inline packages/jsxstyle/preact/jsxstyle-preact.tsx]: Does this work? Using `React` types in `preact`?

File: packages/jsxstyle/preact/jsxstyle-preact.tsx [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  mediaQueries?: { [key: string]: string };
+  props?: { [key: string]: any };
+  style?: string;
+  children: (JSX.Element | JSX.Element[] | string)[];
```
Comment (steadicat) [inline packages/jsxstyle/preact/jsxstyle-preact.tsx]: Can't children also be `null` or `undefined` or `false` or an array of any of the above? Does `preact` not have a `Children` type you can use?

Comment (meyer) [reply packages/jsxstyle/preact/jsxstyle-preact.tsx]: I haven’t spent more than 30 minutes on the preact version so I’m not 10)% sure. I definitely hope preact has a Children type. I pulled this from a `children` key in preact’s type definition file.

File: packages/jsxstyle/preact/jsxstyle-preact.tsx (hunk 19 above)
Comment (meyer) [reply packages/jsxstyle/preact/jsxstyle-preact.tsx]: Nope this is an error. I pasted a bit too much.

File: rollup.config.js (hunk 17 above)
Comment (meyer) [reply rollup.config.js]: Yeah, in this case, Babel wasn’t doing anything, since TypeScript handles JSX, `Object.assign` (at least with JSX spreads), and ES6 classes.

File: packages/jsxstyle/jsxstyle.tsx (hunk 18 above)
Comment (meyer) [reply packages/jsxstyle/jsxstyle.tsx]: @steadicat I can dig it. The only thing I’m not sure about is what I pass to `JsxstyleProps` when I specify it as a type. For example:
```typescript
constructor(props: JsxstyleProps<???>) { // <--- what goes there?
  super(props);
  this.component = props.component || tagName;
  this.className = cache.getClassName(props, props.className);
}```

Sorry, this is probably basic stuff. Still learning this TypeScript biz.

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 13 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Yeah, refactoring is the way to go I think. I was trying to refactor as little as possible to get a first version out ASAP, but in this case it’s quick and easy.

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 12 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Yeah, I didn’t use an optional param because it didn’t seem to change anything (probably because I haven’t hooked up tslint yet). @steadicat mentioned function overloads earlier (https://github.com/smyte/jsxstyle/pull/89#discussion_r151855318). All these options seem kinda gross, but function overloads seem the least gross right now.

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 10 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: @steadicat does this look right?

```typescript
type InsertRuleCallback = (rule: string, props?: {}) => boolean | void;
```

@appsforartists for now, in order to get this PR landed, I’m going to leave this as the not-so-great generic props type, but I’d love a PR to improve this if you have the time. I’m reaching the limits of what I can google regarding TypeScript 😅

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 11 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: ooh, this is neat.

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 15 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: What type would I assign to `styleKeyObj` in that case? I’m not sure how to initialise an empty object with a type in TypeScript without making all the keys optional (which seems wrong to me? idk tho).

**Comment by meyer** [comment]
I gave this PR a bit of thought on my bike ride home. I’d prefer to land it in an imperfect but permissive state—i.e., not breaking everyone‘s builds—than to try to perfect TypeScript support in one PR. Future commits and PRs can make the generic type definitions more specific and fix my noob TypeScript code.

If you all think that’s a bad idea, do let me know. Otherwise I’ll aim to have this landed tomorrow or Wednesday, once I’ve got `jsxstyle/preact` in good shape.

**Comment by meyer** [comment]
Also, that way you all can get commit credit for your contributions! I feel bad copying and pasting your code into my editor 😁

File: packages/jsxstyle/jsxstyle.tsx (hunk 18 above)
Comment (steadicat) [reply packages/jsxstyle/jsxstyle.tsx]: The generic param to `JsxstyleProps` should bubble all the way up to the caller. So, e.g., you’d define `Block` like this:

```ts
class Block<InnerComponentProps> extends React.Component<JsxstyleProps<InnerComponentProps>> {
  constructor(props: JsxstyleProps<InnerComponentProps>) {
    ...
  }
  ...
}
```

This way each call site determines which props are allowed. Ideally `InnerComponentProps` will be inferred by the value of the `component` prop. Should happen magically with the type definition I gave you above. Makes sense?

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 12 above)
Comment (appsforartists) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Personally, I think having a `GetClassNameForKey` type is the most semantic and least gross.  It's a bit more boilerplate, but it expresses your intent clearly (you have a callback that passes both arguments, you want to set it to one that only needs one argument, and you use casting to tell TS that the arity mismatch is OK.)

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 15 above)
Comment (appsforartists) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: I'm not sure if you would need to set it to anything.  I think the type could be inferred in the function body and explicitly set at the function boundary.

```typescript
export type StyleKey = {
  styles: string,
  mediaQuery?: string,
  pseudoclass?: string,
  pseudoelement?: string,
};

export default function getStyleKeysForProps(props: any, pretty = false): Dict<StyleKey> {
  const result = {};
  // build result
  return result
}
```

Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: I was getting this error when I tried that earlier:
```
Type '{}' is not assignable to type 'StyleKeyObject | null'.
```

Comment (appsforartists) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: Maybe I was too optimistic about what it could infer.  I was sleepy and on a plane when I wrote that (and I'm still sleepy and just-off a plane), so I hadn't actually tested it.

In that case, probably:

```typescript
export type StyleKey = {
  styles: string,
  mediaQuery?: string,
  pseudoclass?: string,
  pseudoelement?: string,
};

export default function getStyleKeysForProps(props: any, pretty = false): Dict<StyleKey> {
  const result: Dict<StyleKey> = {};
  // build result
  return result
}
```
You don't even have to `Partial<>` it, because a `Dict` can have no members and still be a `Dict`.

On a related note, I think you're always supposed to give explicit types to the arguments and return value of any exported function.  I know that's true in Flow, but I only presume so in TS.  It's supposed to simplify the amount of work the typechecker has to do if you annotate at the boundaries, which makes type-checking easier to parallelize.

File: packages/jsxstyle/jsxstyle.tsx (hunk 18 above)
Comment (meyer) [reply packages/jsxstyle/jsxstyle.tsx]: Ahhhh yep that makes sense. This is cool!

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 15 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: Heh, no problem. Yeah, that makes sense to use an explicit return type. I wasn’t wild about the inline multiline type definition anyway. Much nicer to have it top-level.

File: packages/jsxstyle/jsxstyle.tsx (hunk 18 above)
Comment (meyer) [reply packages/jsxstyle/jsxstyle.tsx]: @steadicat does it make sense for `AnyComponent`’s props to extend `ComponentPropProps`? I’d like TypeScript to complain if a component specified as the `component` prop doesn’t accept the `className` prop (and perhaps the `style` prop as well).

Example:

```typescript
export interface ComponentPropProps {
  className?: any;
  style?: CSSProperties;
  [key: string]: any;
}

export type AnyComponent<Props extends ComponentPropProps> =
  | keyof JSX.IntrinsicElements
  | React.ComponentClass<Props>
  | React.SFC<Props>;

// ...
```

I think this does what I expect, but I might be wrong. React might already set `style` to `React.CSSProperties` and `className` to something like `any`? I’m not 100% sure.

Comment (steadicat) [reply packages/jsxstyle/jsxstyle.tsx]: Yes! You got this! The index accessor on `ComponentPropProps` shouldn't be needed. Just specify the props you need the component to accept. I believe `className` is something like `string | null | undefined | false`. VSCode is great when you're discovering this stuff. You can hover over anything to see the inferred type, and jump to definition to see the full type declaration. Also, nit: the type of `style?: CSSProperties` should probably be the native React one, not the Jsxstyle kind.

**Comment by meyer** [comment]
Preact version is a bit gross but it builds. It’s shippable (imo) but I’d love for someone who uses Preact to improve it at some point in the future.

I think this is ready to land. Lay your final feedback upon me.

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts [hunk 21]
```diff
@@ -35,7 +35,19 @@ const specialCaseProps = {
   mediaQueries: true,
 };
 
-export default function getStyleKeysForProps(props: any, pretty = false) {
+export type StyleKeyObj = {
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: This is a `Dict`!

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts [hunk 22]
```diff
@@ -52,14 +64,7 @@ export default function getStyleKeysForProps(props: any, pretty = false) {
...
   // return value
-  const styleKeyObj: {
-    [key: string]: {
-      styles: string;
-      mediaQuery?: string;
-      pseudoclass?: string;
-      pseudoelement?: string;
-    };
-  } & { classNameKey: string } = <any>{};
+  const styleKeyObj: StyleKeyObj = <any>{};
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: Can you cast this to `StyleKeyObj` instead of `any`?

(I still think you should be using `as` for casting, but that's a style thing.)

File: packages/jsxstyle-utils/src/getStyleCache.ts [hunk 23]
```diff
@@ -16,7 +17,8 @@ function alreadyInjected() {
   );
 }
 
-function getStringHash(key: string, props: {}) {
+function getStringHash(key: string, props: {}): string;
+function getStringHash(key: string) {
   return '_' + stringHash(key).toString(36);
 }
```
Comment (appsforartists) [inline packages/jsxstyle-utils/src/getStyleCache.ts]: `getStyleCache` is missing a return type annotation

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 21 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: 🙏 thx

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 23 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Yeah, it’s inferred in the second case.

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 22 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: Yep, that works. Thanks!

File: packages/jsxstyle-utils/src/getStyleCache.ts (hunk 23 above)
Comment (meyer) [reply packages/jsxstyle-utils/src/getStyleCache.ts]: Thought about this a bit more. It’s probably best to be explicit.

File: packages/jsxstyle-utils/src/getStyleKeysForProps.ts (hunk 22 above)
Comment (appsforartists) [reply packages/jsxstyle-utils/src/getStyleKeysForProps.ts]: I might even call it `StyleKeyDict` to be consistent with your naming.

**Comment by appsforartists** [comment]
:tada:!

**Comment by meyer** [comment]
Published as `jsxstyle@2.0.5` and `jsxstyle-utils@2.0.3`.
